pallets/free-calls/src/lib.rs

```diff
@@ -0,0 +1,430 @@
+                Self::can_make_free_call(&sender, ShouldUpdateConsumerStats::YES) {
+                // Add the current weight for the boxed call
+                actual_weight = actual_weight.saturating_add(call.get_dispatch_info().weight);
+
+                // Dispatch the call
+                let result = call.dispatch(origin);
+
+                // Deposit an event with the result
+                Self::deposit_event(Event::FreeCallResult(
+                    sender,
+                    result.map(|_| ()).map_err(|e| e.error),
+                ));
+            }
+
+            Ok(PostDispatchInfo {
+                actual_weight: Some(actual_weight),
+                pays_fee: Pays::No,
+            })
+        }
+    }
+
+    struct Window<T: Config> {
+        consumer: T::AccountId,
+        config_index: WindowType,
+        config: WindowConfig<T::BlockNumber>,
+        timeline_index: T::BlockNumber,
+        stats: ConsumerStats<T::BlockNumber>,
+        can_have_free_call: bool,
+    }
+
+    impl<T: Config> Window<T> {
+        // TODO: refactor this into more lightweight version??
+        fn build(
+            consumer: T::AccountId,
+            quota: NumberOfCalls,
+            current_block: T::BlockNumber,
+            config_index: WindowType,
+            config: WindowConfig<T::BlockNumber>,
+            window_stats: Option<ConsumerStats<T::BlockNumber>>,
+        ) -> Self {
+            let timeline_index = current_block / config.period;
+
+            let reset_stats = || ConsumerStats::new(timeline_index);
+
+            let mut stats = window_stats.unwrap_or_else(reset_stats);
+
+            if stats.timeline_index < timeline_index {
+                stats = reset_stats();
+            }
+
+            let can_be_called = stats.used_calls < max(1, quota / config.quota_ratio);
+
+            Window {
+                consumer: consumer.clone(),
+                config_index,
+                config,
+                timeline_index,
+                stats,
+                can_have_free_call: can_be_called,
+            }
+        }
+
+        fn increment_window_stats(&mut self) {
+            self.stats.used_calls = self.stats.used_calls.saturating_add(1);
+            <WindowStatsByConsumer<T>>::insert(
+                self.consumer.clone(),
+                self.config_index,
+                self.stats.clone(),
+            );
+        }
+    }
+
+    pub enum ShouldUpdateConsumerStats {
+        YES,
+        NO,
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// Determine if `consumer` can have a free call and optionally update its window usage.
+        ///
+        /// Window usage for the caller `consumer` will only update if there is a quota and all of the
+        /// previous window usages doesn't exceed the defined windows config.
+        pub fn can_make_free_call(consumer: &T::AccountId, should_update_consumer_stats: ShouldUpdateConsumerStats) -> bool {
```

Comment:
i generally do not like combining a check function with also a state modification function, but that is just up to you

Reply:
Yeah I planned to refactor this, which will make it easier for tests, and also it will be less complex.